### ISSUE TITLE
Is this vulnerable to the presigned GetCallerIdentity issue (CVE-2020-16250) impacting Hashicorp Vault?

### ISSUE THREAD

**Issue by 0xdabbad00** [kubernetes-sigs/aws-iam-authenticator#340]
Google's Project Zero posted this issue today with Hashicorp Vault: https://googleprojectzero.blogspot.com/2020/10/enter-the-vault-auth-issues-hashicorp-vault.html

Your README describes a similar mechanism and states "This mechanism is borrowed with a few changes from Vault."   I've not dug into the code, but I'm suspicious that you may be vulnerable to something similar.

**Comment by aidansteele** [comment]
Looks like it's not vulnerable to the same issue https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L466-L468

**Comment by mattmoyer** [comment]
tl;dr no, we're safe

Some caveats:
- I originally wrote some of the code in question, but I'm not the most active maintainer recently.
- This is not an official security opinion of my employer.
- I'm not a full time security researcher, but I do have some security background and was feeling particularly paranoid when I copied this approach from Vault originally 😄 .

I think aws-iam-authenticator is not vulnerable to the techniques demonstrated in the excellent Project Zero writeup:

1. We have stricter validation for the signed request. In aws-iam-authenticator the request is always just a single URL, parsed by `net/url` which is pretty well hardened against URL parser tricks. We don't allow a request body at all, force the upstream STS request to [always be a `GET`](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L500), [allow only certain known URL parameters](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L134-L144), and as @aidansteele mentioned, we [check that `action` is always `GetCallerIdentity`](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L466-L468).

   Together, I think these should always prevent the root bug that P0 demonstrated in Vault, and I'm cautiously optimistic that this code is solid.

1. As of v0.5.2 (released today) the `Host` validation was [switched](https://github.com/kubernetes-sigs/aws-iam-authenticator/pull/341) from a regex to one [based on a list of STS endpoints in each AWS partition](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L407-L413) I'm not aware of any known vulnerabilities in the previous validation here, but the new approach seems very robust.

1. We do not rely on the Go XML parser libraries, instead we always make the STS request with [`Accept: application/json`](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L502) and use `encoding/json` to [unmarshal the response body](https://github.com/kubernetes-sigs/aws-iam-authenticator/blob/292b9b82df69b87af962b92485b254d9f4b10f00/pkg/token/token.go#L523-L527). This was mostly accidental, but I would expect the `encoding/json` parser to be more strict about the input than `encoding/xml` seems to be. I don't know that this is foolproof, but it seems like a bit more rigid mechanism.

Again, this is just my thoughts after an initial triage, but I'm comfortable closing the issue for now.

I'd encourage anyone with more idea for further hardening to please share here. If you think you've found a way to exploit this or any other vulnerability in a Kubernetes component, please follow the [disclosure procedure](https://github.com/kubernetes-sigs/aws-iam-authenticator/security/policy).

Thanks for sharing!

/label security
/close

**Comment by 0xdabbad00** [comment]
Felix Wilhelm, who found the original issue with Vault, was able to find an issue with this code, although a different one than the Vault project had: https://bugs.chromium.org/p/project-zero/issues/detail?id=2066
